pandevice/base.py

```diff
@@ -1367,18 +1367,73 @@ def _convert_var(cls, value, vartype):
         elif vartype == "bool":
             return yesno(value)
 
-    def _set_reference(self, reference_name, reference_type, reference_var, exclusive, refresh, update, running_config, *args, **kwargs):
+    def _set_reference(self, reference_name, reference_type, reference_var,
+                       exclusive, refresh, update, running_config,
+                       return_boolean, name_only, **kwargs):
         """Used by helper methods to set references between objects
 
         For example, set_zone() would set the zone for an interface by creating a reference from
         the zone to the interface. If the desired reference already exists then nothing happens.
 
         """
-        device = self.nearest_pandevice()
+        parent = None
+        update_needed = False
+
         if refresh:
-            allobjects = reference_type.refreshall(device, running_config=running_config)
+            from pandevice.firewall import Firewall
```

Comment:
Can you use comments here to document why this new refresh algorithm is necessary and, at a high level, what it does?